static/js/narrow.js

```diff
@@ -272,6 +253,142 @@ export function activate(raw_operators, opts) {
         id_info.target_id = Number.parseInt(filter.operands("id")[0], 10);
     }
 
+    // Narrow with near / id operator. There are two possibilities:
+    // * The user is clicking a permanent link to a conversation, in which
+    //   case we want to look up the anchor message and see if it has moved.
+    // * The user did a search for something like stream:foo topic:bar near:1
+    //   (or some other ID that is not an actual message in the topic).
+    //
+    // We attempt the match the stream and topic with that of the
+    // message in case the message was moved after the link was
+    // created. This ensures near / id links work and will redirect
+    // correctly if the topic was moved (including being resolved).
+    if (id_info.target_id && filter.has_operator("stream") && filter.has_operator("topic")) {
+        const target_message = message_store.get(id_info.target_id);
+        const adjusted_operators = [];
+        let operators_changed = false;
+
+        function adjusted_operators_if_moved(operators, message) {
+            for (const operator of operators) {
+                const adjusted_operator = {...operator};
+                if (
+                    operator.operator === "stream" &&
+                    operator.operand !== message.display_recipient
```

Comment:
Yeah, I think a `util.lower_same` comparison would be most correct here. It's likely hard to make this actually come up because I suspect they are both sourced from the current stream's name from `stream_data.js`, but it's better to not have to think about that plumbing question.

Reply:
Updated in latest update.